tscat/__init__.py

```diff
@@ -46,61 +59,201 @@ def variable_attributes_as_dict(self):
+
+        if key != '_in_ctor' and not self._in_ctor:
+            # TODO use backend().add
+            if key in self._fixed_keys:
+                setattr(self._backend_entity, key, value)
+            elif _valid_key.match(key):
+                self._backend_entity[key] = value
+
+    def __delattr__(self, key):
+        super(_BackendBasedEntity, self).__delattr__(key)
+
+        # only allow deletion of attributes
+        if key in self._fixed_keys:
+            raise IndexError('Fixed keys cannot be deleted.')
+        if _valid_key.match(key):
+            # backend().del
+            del self._backend_entity[key]
 
+    def __eq__(self, o):
+        if sorted(filter(_valid_key.match, self.__dict__.keys())) != \
+            sorted(filter(_valid_key.match, o.__dict__.keys())):
+            return False
+
+        for k in sorted(filter(_valid_key.match, self.__dict__.keys())):
+            if self.__dict__[k] != o.__dict__[k]:
+                return False
+        return True
+
+
+@typechecked
 class Event(_BackendBasedEntity):
-    _fixed_keys = ['start', 'stop', 'author', 'uuid']
+    _fixed_keys = ['start', 'stop', 'author', 'uuid', 'tags', 'products']
 
-    def __init__(self, start: dt.datetime, stop: dt.datetime, author: str, uuid=None, **kwargs):
+    def __init__(self, start: dt.datetime, stop: dt.datetime,
+                 author: str,
+                 uuid: str = None,
+                 tags: Iterable[str] = [],
+                 products: Iterable[str] = [],
+                 _insert: bool = True,
+                 **kwargs):
+        self._in_ctor = True
         super().__init__()
 
-        if start >= stop:
-            raise ValueError("start-datetime has to be younger than end-datetime")
-
         self.start = start
         self.stop = stop
         self.author = author
+        self.tags = list(tags)
+        self.products = list(products)
 
         if not uuid:
             self.uuid = str(uuid4())
         else:
-            UUID(uuid, version=4)  # check whether it is a valid uuid - throws ValueError if not
             self.uuid = uuid
 
-        _create_attributes(self, kwargs)
+        _verify_attribute_names(kwargs)
+        self.__dict__.update(kwargs)
+
+        if _insert:
+            self._backend_entity = backend().add_or_update_event({
+                'start': self.start,
+                'stop': self.stop,
+                'author': self.author,
+                'uuid': self.uuid,
+                'tags': self.tags,
+                'products': self.products,
+                'attributes': kwargs,
+            })
+
+        self._in_ctor = False
+
+    def __setattr__(self, key, value):
+        if key == 'uuid':
+            UUID(value, version=4)  # throws an exception if not valid
+        elif key == 'start' and hasattr(self, 'stop'):
+            if value >= self.stop:
+                raise ValueError("start-datetime has to be younger than end-datetime")
```

Comment:
younger isn't appropriate since start-date isn't a person and there seems to be one or two copy paste mistakes:
- mixing end and stop
- message bellow is even worse
It's also weird to mix start/end with datetime which is more an object type than a thing.

Reply:
Yes, agreed error messages needs to be fixed.

What do you by 'weird mixing start/end with datetime'?

Reply:
I changed the messages to use before/after and fixed end vs. stop-date